pkg/cable/iptun/iptun.go

```diff
@@ -0,0 +1,356 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptun
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/cable"
+	"github.com/submariner-io/submariner/pkg/cable/cableutils"
+	"github.com/submariner-io/submariner/pkg/natdiscovery"
+	netlinkAPI "github.com/submariner-io/submariner/pkg/netlink"
+	"github.com/submariner-io/submariner/pkg/routeagent_driver/cni"
+	"github.com/submariner-io/submariner/pkg/types"
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog"
+)
+
+const (
+	IPTunIface         = "ipip-tunnel"
+	IPTunOverhead      = 20
+	IPTunNetworkPrefix = 242
+	CableDriverName    = "iptun"
+	TableID            = 100
+)
+
+type Operation int
+
+const (
+	Add Operation = iota
+	Delete
+	Flush
+)
+
+type iptun struct {
+	localEndpoint types.SubmarinerEndpoint
+	localCluster  types.SubmarinerCluster
+	connections   []v1.Connection
+	mutex         sync.Mutex
+	ipTunIface    *ipTunIface
+	netLink       netlinkAPI.Interface
+}
+
+type ipTunIface struct {
+	activeEndpointHostname string
+	link                   *netlink.Iptun
+	iptunIP                net.IP
+}
+
+type iptunAttributes struct {
+	name     string
+	pmtuDisc uint8
+	local    net.IP
+	remote   net.IP
+	mtu      int
+}
+
+func init() {
+	cable.AddDriver(CableDriverName, NewDriver)
+}
+
+func NewDriver(localEndpoint *types.SubmarinerEndpoint, localCluster *types.SubmarinerCluster) (cable.Driver, error) {
+	// We'll panic if localEndpoint or localCluster are nil, this is intentional
+	ipip := iptun{
+		localEndpoint: *localEndpoint,
+		netLink:       netlinkAPI.New(),
+		localCluster:  *localCluster,
+	}
+
+	// Don't create an ipip-tunnel interface till you have the remote endpoint info.
+	return &ipip, nil
+}
+
+func (ipip *iptun) createIPTunInterface(remoteIP net.IP) error {
+	ipAddr := ipip.localEndpoint.Spec.PrivateIP
+
+	iptunIP, err := cableutils.GetTunnelEndPointIPAddress(ipAddr, IPTunNetworkPrefix)
+	if err != nil {
+		return errors.Wrapf(err, "failed to derive the iptun IP for %s", ipAddr)
+	}
+
+	defaultHostIface, err := netlinkAPI.GetDefaultGatewayInterface()
+	if err != nil {
+		return errors.Wrapf(err, "Unable to find the default interface on host: %s",
+			ipip.localEndpoint.Spec.Hostname)
+	}
+
+	// Derive the MTU based on the default outgoing interface.
+	iptunMtu := defaultHostIface.MTU - IPTunOverhead
+
+	attrs := &iptunAttributes{
+		name:     IPTunIface,
+		local:    net.ParseIP(ipAddr),
+		remote:   remoteIP,
+		mtu:      iptunMtu,
+		pmtuDisc: 1,
+	}
+
+	ipip.ipTunIface, err = newIPTunIface(attrs, ipip.localEndpoint.Spec.Hostname)
```

Comment:
So this is called per remote endpoint and creates an `ipTunIface` instance with endpoint-specific info (ie `remoteIP`) but it is stored in a shared field. This seems problematic unless I'm missing something.